components/ledger/pkg/machine/script/compiler/compiler.go

```diff
@@ -78,38 +78,58 @@ func (p *parseVisitor) VisitVariable(c parser.IVariableContext, push bool) (core
 func (p *parseVisitor) VisitExpr(c parser.IExpressionContext, push bool) (core.Type, *core.Address, *CompileError) {
 	switch c := c.(type) {
 	case *parser.ExprAddSubContext:
-		ty, _, err := p.VisitExpr(c.GetLhs(), push)
+		lhsType, lhsAddr, err := p.VisitExpr(c.GetLhs(), push)
 		if err != nil {
 			return 0, nil, err
 		}
-		if ty != core.TypeNumber {
-			return 0, nil, LogicError(c, errors.New("tried to do arithmetic with wrong type"))
-		}
-		ty, _, err = p.VisitExpr(c.GetRhs(), push)
-		if err != nil {
-			return 0, nil, err
-		}
-		if ty != core.TypeNumber {
-			return 0, nil, LogicError(c, errors.New("tried to do arithmetic with wrong type"))
-		}
-		if push {
-			switch c.GetOp().GetTokenType() {
-			case parser.NumScriptLexerOP_ADD:
-				p.AppendInstruction(program.OP_IADD)
-			case parser.NumScriptLexerOP_SUB:
-				p.AppendInstruction(program.OP_ISUB)
+		switch lhsType {
+		case core.TypeNumber:
+			rhsType, _, err := p.VisitExpr(c.GetRhs(), push)
+			if err != nil {
+				return 0, nil, err
+			}
+			if rhsType != core.TypeNumber {
+				return 0, nil, LogicError(c, fmt.Errorf(
+					"tried to do an arithmetic operation with incompatible left and right-hand side operand types: %s and %s",
+					lhsType, rhsType))
+			}
+			if push {
+				switch c.GetOp().GetTokenType() {
+				case parser.NumScriptLexerOP_ADD:
+					p.AppendInstruction(program.OP_IADD)
+				case parser.NumScriptLexerOP_SUB:
+					p.AppendInstruction(program.OP_ISUB)
+				}
+			}
+			return core.TypeNumber, nil, nil
+		case core.TypeMonetary:
+			rhsType, _, err := p.VisitExpr(c.GetRhs(), push)
+			if err != nil {
+				return 0, nil, err
 			}
+			if rhsType != core.TypeMonetary {
+				return 0, nil, LogicError(c, fmt.Errorf(
+					"tried to do an arithmetic operation with incompatible left and right-hand side operand types: %s and %s",
+					lhsType, rhsType))
+			}
+			if push {
+				switch c.GetOp().GetTokenType() {
+				case parser.NumScriptLexerOP_ADD:
+					p.AppendInstruction(program.OP_MONETARY_ADD)
+				case parser.NumScriptLexerOP_SUB:
+					p.AppendInstruction(program.OP_MONETARY_SUB)
+				}
+			}
+			return core.TypeMonetary, lhsAddr, nil
```

Comment:
I think this is misleading. I know we use the left monetary's address to gain access to its asset, but it isn't clearly expressed in this API (it looks like the second return value will be the current expr's value even though it isn't).  This is why in my draft I made a separate struct for VisitExpr's return, so that it's clearer that `asset_source` is *not* the actual address (though asset_source isn't a very good name because of the other meaning of "source" 😅)